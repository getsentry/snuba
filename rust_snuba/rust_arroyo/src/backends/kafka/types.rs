use rdkafka::message::OwnedHeaders;

#[derive(Clone)]
pub struct KafkaPayload {
    pub key: Option<Vec<u8>>,
    pub headers: Option<OwnedHeaders>,
    pub payload: Option<Vec<u8>>,
}
