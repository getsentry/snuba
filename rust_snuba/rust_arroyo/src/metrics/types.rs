use core::fmt::{self, Display};
use std::time::Duration;

/// The Type of a Metric.
///
/// Counters, Gauges and Distributions are supported,
/// with more types to be added later.
#[non_exhaustive]
#[derive(Debug)]
pub enum MetricType {
    /// A counter metric, using the StatsD `c` type.
    Counter,
    /// A gauge metric, using the StatsD `g` type.
    Gauge,
    /// A timer metric, using the StatsD `ms` type.
    Timer,
    // Distribution,
    // Meter,
    // Histogram,
    // Set,
}

impl MetricType {
    /// Returns the StatsD metrics type.
    pub fn as_str(&self) -> &str {
        match self {
            MetricType::Counter => "c",
            MetricType::Gauge => "g",
            MetricType::Timer => "ms",
        }
    }
}

impl Display for MetricType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(self.as_str())
    }
}

/// A Metric Value.
///
/// This supports various numeric values for now, but might gain support for
/// `Duration` and other types later on.
#[non_exhaustive]
#[derive(Debug)]
pub enum MetricValue {
    /// A signed value.
    I64(i64),
    /// An unsigned value.
    U64(u64),
    /// A floating-point value.
    F64(f64),
    /// A [`Duration`] value.
    Duration(Duration),
}

impl Display for MetricValue {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            MetricValue::I64(v) => v.fmt(f),
            MetricValue::U64(v) => v.fmt(f),
            MetricValue::F64(v) => v.fmt(f),
            MetricValue::Duration(d) => d.as_millis().fmt(f),
        }
    }
}

macro_rules! into_metric_value {
    ($($from:ident),+ => $variant:ident) => {
        $(
            impl From<$from> for MetricValue {
                #[inline(always)]
                fn from(f: $from) -> Self {
                    Self::$variant(f.into())
                }
            }
        )+
    };
}

into_metric_value!(i8, i16, i32, i64 => I64);
into_metric_value!(u8, u16, u32, u64 => U64);
into_metric_value!(f32, f64 => F64);
into_metric_value!(Duration => Duration);

/// An alias for a list of Metric tags.
pub type MetricTags<'a> = &'a [(Option<&'a dyn Display>, &'a dyn Display)];

/// A fully types Metric.
///
/// Most importantly, the metric has a [`ty`](MetricType), a `key` and a [`value`](MetricValue).
/// It can also have a list of [`tags`](MetricTags).
///
/// This struct might change in the future, and one should construct it via
/// the [`metric!`](crate::metric) macro instead.
pub struct Metric<'a> {
    /// The key, or name, of the metric.
    pub key: &'a dyn Display,
    /// The type of metric.
    pub ty: MetricType,

    /// A list of tags for this metric.
    pub tags: MetricTags<'a>,
    /// The metrics value.
    pub value: MetricValue,

    #[doc(hidden)]
    pub __private: (),
}
