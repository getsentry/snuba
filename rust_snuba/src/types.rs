use std::cmp::min;
use std::collections::BTreeMap;

use crate::strategies::clickhouse::batch::HttpBatch;

use chrono::{DateTime, Utc};
use rust_arroyo::backends::kafka::types::KafkaPayload;
use rust_arroyo::timer;
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, PartialEq)]
pub struct CommitLogEntry {
    pub offset: u64,
    pub orig_message_ts: DateTime<Utc>,
    pub received_p99: Vec<DateTime<Utc>>,
}

#[derive(Default, Debug, Clone)]
pub struct CommitLogOffsets(pub BTreeMap<u16, CommitLogEntry>);

impl CommitLogOffsets {
    fn merge(&mut self, other: CommitLogOffsets) {
        for (partition, other_entry) in other.0 {
            self.0
                .entry(partition)
                .and_modify(|entry| {
                    entry.offset = other_entry.offset;
                    entry.orig_message_ts = other_entry.orig_message_ts;
                    entry.received_p99.extend(&other_entry.received_p99);
                })
                .or_insert(other_entry);
        }
    }
}

#[derive(Clone, Debug, PartialEq, Default)]
pub struct CogsData {
    pub data: BTreeMap<String, u64>, // app_feature: bytes_len
}

impl CogsData {
    fn merge(&mut self, other: CogsData) {
        for (k, v) in other.data {
            self.data
                .entry(k)
                .and_modify(|curr| *curr += v)
                .or_insert(v);
        }
    }
}

#[derive(Debug, Clone)]
struct LatencyRecorder {
    sum_timestamps: f64,
    earliest_timestamp: u64,
    num_values: usize,
}

impl Default for LatencyRecorder {
    fn default() -> Self {
        LatencyRecorder {
            sum_timestamps: 0.0,
            earliest_timestamp: u64::MAX,
            num_values: 0,
        }
    }
}

impl From<DateTime<Utc>> for LatencyRecorder {
    fn from(value: DateTime<Utc>) -> Self {
        let value = value.timestamp_millis();
        LatencyRecorder {
            sum_timestamps: value as f64,
            earliest_timestamp: value as u64,
            num_values: 1,
        }
    }
}

impl LatencyRecorder {
    fn merge(&mut self, other: Self) {
        self.sum_timestamps += other.sum_timestamps;
        self.earliest_timestamp = min(self.earliest_timestamp, other.earliest_timestamp);
        self.num_values += other.num_values;
    }

    fn max_value_ms(&self, write_time: DateTime<Utc>) -> u64 {
        let write_time = write_time.timestamp_millis() as u64;
        write_time.saturating_sub(self.earliest_timestamp)
    }

    fn avg_value_ms(&self, write_time: DateTime<Utc>) -> u64 {
        let write_time = write_time.timestamp_millis() as u64;
        (write_time as f64 - (self.sum_timestamps / self.num_values as f64)) as u64
    }

    fn send_metric(&self, write_time: DateTime<Utc>, metric_name: &str) {
        if self.num_values == 0 {
            return;
        }

        timer!(
            format_args!("insertions.max_{}_ms", metric_name),
            self.max_value_ms(write_time)
        );

        timer!(
            format_args!("insertions.{}_ms", metric_name),
            self.avg_value_ms(write_time),
        );
    }
}

#[derive(Clone, Debug, PartialEq)]
pub struct ReplacementData {
    pub key: Vec<u8>, // Project id
    // Replacement message bytes. Newline-delimited series of json payloads
    pub value: Vec<u8>,
}

impl From<ReplacementData> for KafkaPayload {
    fn from(value: ReplacementData) -> KafkaPayload {
        KafkaPayload::new(Some(value.key), None, Some(value.value))
    }
}

impl From<KafkaPayload> for ReplacementData {
    fn from(value: KafkaPayload) -> ReplacementData {
        ReplacementData {
            key: value.key().unwrap().clone(),
            value: value.payload().unwrap().clone(),
        }
    }
}

#[derive(Clone, Debug, PartialEq)]
pub enum InsertOrReplacement<T> {
    Insert(T),
    Replacement(ReplacementData),
}

/// The return value of message processors.
#[derive(Clone, Debug, Default, PartialEq)]
pub struct InsertBatch {
    pub rows: RowData,
    pub origin_timestamp: Option<DateTime<Utc>>,
    pub sentry_received_timestamp: Option<DateTime<Utc>>,
    pub cogs_data: Option<CogsData>,
}

impl InsertBatch {
    pub fn from_rows<T>(
        rows: impl IntoIterator<Item = T>,
        origin_timestamp: Option<DateTime<Utc>>,
    ) -> anyhow::Result<Self>
    where
        T: Serialize,
    {
        let rows = RowData::from_rows(rows)?;
        Ok(Self {
            rows,
            origin_timestamp,
            sentry_received_timestamp: None,
            cogs_data: None,
        })
    }

    /// In case the processing function wants to skip the message, we return an empty batch.
    /// But instead of having the caller send an empty batch, lets make an explicit api for
    /// skipping. This way we can change the implementation later if we want to. Skipping ensures
    /// that the message is committed but not processed.
    pub fn skip() -> Self {
        Self::default()
    }
}

#[derive(Clone, Debug, Default)]
pub struct BytesInsertBatch<R> {
    rows: R,

    /// when the message was inserted into the snuba topic
    ///
    /// In Python this aggregate value is not explicitly tracked on BytesInsertBatch, as the batch
    /// type contains a list of the original kafka messages which all contain the individual
    /// timestamp values in metadata.
    message_timestamp: LatencyRecorder,

    /// when the event was received by Relay
    origin_timestamp: LatencyRecorder,

    /// when it was received by the ingest consumer in Sentry
    ///
    /// May not be recorded for some datasets. This is specifically used for metrics datasets, where
    /// this represents the latency from ingest-metrics (i.e. before the metrics indexer) to
    /// insertion into clickhouse
    sentry_received_timestamp: LatencyRecorder,

    // For each partition we store the offset and timestamp to be produced to the commit log
    commit_log_offsets: CommitLogOffsets,

    cogs_data: CogsData,
}

impl<R> BytesInsertBatch<R> {
    pub fn new(
        rows: R,
        message_timestamp: Option<DateTime<Utc>>,
        origin_timestamp: Option<DateTime<Utc>>,
        sentry_received_timestamp: Option<DateTime<Utc>>,
        commit_log_offsets: CommitLogOffsets,
        cogs_data: CogsData,
    ) -> Self {
        BytesInsertBatch {
            rows,
            message_timestamp: message_timestamp
                .map(LatencyRecorder::from)
                .unwrap_or_default(),
            origin_timestamp: origin_timestamp
                .map(LatencyRecorder::from)
                .unwrap_or_default(),
            sentry_received_timestamp: sentry_received_timestamp
                .map(LatencyRecorder::from)
                .unwrap_or_default(),
            commit_log_offsets,
            cogs_data,
        }
    }

    pub fn commit_log_offsets(&self) -> &CommitLogOffsets {
        &self.commit_log_offsets
    }

    pub fn cogs_data(&self) -> &CogsData {
        &self.cogs_data
    }

    pub fn take(self) -> (R, BytesInsertBatch<()>) {
        let new = BytesInsertBatch {
            rows: (),
            message_timestamp: self.message_timestamp,
            origin_timestamp: self.origin_timestamp,
            sentry_received_timestamp: self.sentry_received_timestamp,
            commit_log_offsets: self.commit_log_offsets,
            cogs_data: self.cogs_data,
        };

        (self.rows, new)
    }

    pub fn record_message_latency(&self) {
        let write_time = Utc::now();

        self.message_timestamp.send_metric(write_time, "latency");
        self.origin_timestamp
            .send_metric(write_time, "end_to_end_latency");
        self.sentry_received_timestamp
            .send_metric(write_time, "sentry_received_latency");
    }
}

impl BytesInsertBatch<RowData> {
    pub fn len(&self) -> usize {
        self.rows.num_rows
    }
}

impl BytesInsertBatch<HttpBatch> {
    pub fn merge(mut self, other: BytesInsertBatch<RowData>) -> Self {
        self.rows
            .write_rows(&other.rows)
            .expect("failed to write rows to channel");
        self.commit_log_offsets.merge(other.commit_log_offsets);
        self.message_timestamp.merge(other.message_timestamp);
        self.origin_timestamp.merge(other.origin_timestamp);
        self.sentry_received_timestamp
            .merge(other.sentry_received_timestamp);
        self.cogs_data.merge(other.cogs_data);
        self
    }
}

#[derive(Clone, Debug, Deserialize, Serialize, Default, PartialEq)]
pub struct RowData {
    pub encoded_rows: Vec<u8>,
    pub num_rows: usize,
}

impl RowData {
    pub fn from_rows<T>(rows: impl IntoIterator<Item = T>) -> anyhow::Result<Self>
    where
        T: Serialize,
    {
        let mut encoded_rows = Vec::new();
        let mut num_rows = 0;
        for row in rows {
            serde_json::to_writer(&mut encoded_rows, &row)?;
            debug_assert!(encoded_rows.ends_with(b"}"));
            encoded_rows.push(b'\n');
            num_rows += 1;
        }

        Ok(RowData {
            num_rows,
            encoded_rows,
        })
    }

    pub fn from_encoded_rows(rows: Vec<Vec<u8>>) -> Self {
        let mut encoded_rows = Vec::new();
        let mut num_rows = 0;
        for row in rows {
            encoded_rows.extend_from_slice(&row);
            encoded_rows.push(b'\n');
            num_rows += 1;
        }

        RowData {
            encoded_rows,
            num_rows,
        }
    }

    pub fn into_encoded_rows(self) -> Vec<u8> {
        self.encoded_rows
    }
}

#[derive(Clone, Debug)]
pub struct KafkaMessageMetadata {
    pub partition: u16,
    pub offset: u64,
    pub timestamp: DateTime<Utc>,
}

#[cfg(test)]
mod tests {
    use chrono::{TimeZone, Utc};

    use super::*;

    #[test]
    fn test_latency_recorder_basic() {
        // emulate how Reduce works when creating a batch for inserting into clickhouse, because
        // that's where we use it in practice. Reduce starts with a default() accumulator
        let mut accumulator = LatencyRecorder::default();

        // then there's one row being merged into the current batch
        let now = Utc.timestamp_opt(61, 0).unwrap();
        accumulator.merge(LatencyRecorder::from(now));

        // then another row
        let now = Utc.timestamp_opt(63, 0).unwrap();
        accumulator.merge(LatencyRecorder::from(now));

        // finally, when flushing the batch, we record the metric. let's see if LatencyRecorder did
        // the math right:
        assert_eq!(accumulator.num_values, 2);
        let now = Utc.timestamp_opt(65, 0).unwrap();
        assert_eq!(accumulator.max_value_ms(now), 4000);
        assert_eq!(accumulator.avg_value_ms(now), 3000);
    }
}
