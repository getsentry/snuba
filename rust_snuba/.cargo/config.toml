[env]
# Workaround for https://github.com/confluentinc/librdkafka/pull/5012
CMAKE_POLICY_VERSION_MINIMUM = "3.10"
