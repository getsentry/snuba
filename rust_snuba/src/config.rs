use std::collections::HashMap;

use serde::{Deserialize, Deserializer};
use serde_json::Value;

#[derive(Deserialize, Debug)]
#[serde(deny_unknown_fields)]
pub struct ConsumerConfig {
    pub storages: Vec<StorageConfig>,
    pub raw_topic: TopicConfig,
    pub commit_log_topic: Option<TopicConfig>,
    pub replacements_topic: Option<TopicConfig>,
    pub dlq_topic: Option<TopicConfig>,
    pub max_batch_size: usize,
    pub max_batch_time_ms: u64,
    pub env: EnvConfig,
}

pub fn deserialize_broker_config<'de, D>(
    deserializer: D,
) -> Result<HashMap<String, String>, D::Error>
where
    D: Deserializer<'de>,
{
    let data = RawBrokerConfig::deserialize(deserializer)?
        .iter()
        .filter_map(|(k, v)| {
            if v.is_null() {
                None
            } else if v.is_number() {
                // Numeric types are valid in confluent-kafka-python config but not in the Rust library
                Some((k.to_string(), v.as_number().unwrap().to_string()))
            } else if v.is_string() {
                if v.as_str().unwrap().is_empty() {
                    return None;
                }
                Some((k.to_string(), v.as_str().unwrap().to_string()))
            } else {
                panic!("Unsupported type");
            }
        })
        .collect();

    Ok(data)
}

#[derive(Deserialize, Debug)]
pub struct TopicConfig {
    pub physical_topic_name: String,
    pub logical_topic_name: String,
    #[serde(deserialize_with = "deserialize_broker_config")]
    pub broker_config: BrokerConfig,
}

type RawBrokerConfig = HashMap<String, Value>;

pub type BrokerConfig = HashMap<String, String>;

impl ConsumerConfig {
    pub fn load_from_str(payload: &str) -> Result<Self, anyhow::Error> {
        let d: Self = serde_json::from_str(payload)?;
        Ok(d)
    }
}

#[derive(Deserialize, Clone, Debug)]
#[serde(deny_unknown_fields)]
pub struct StorageConfig {
    pub name: String,
    pub clickhouse_table_name: String,
    pub clickhouse_cluster: ClickhouseConfig,
    pub message_processor: MessageProcessorConfig,
}

#[derive(Deserialize, Clone, Debug)]
#[serde(deny_unknown_fields)]
pub struct ClickhouseConfig {
    pub host: String,
    pub port: u16,
    pub http_port: u16,
    pub user: String,
    pub password: String,
    pub database: String,
}

#[derive(Deserialize, Clone, Debug)]
#[serde(deny_unknown_fields)]
pub struct MessageProcessorConfig {
    pub python_class_name: String,
    pub python_module: String,
}

#[derive(Deserialize, Debug)]
#[serde(deny_unknown_fields)]
pub struct EnvConfig {
    pub sentry_dsn: Option<String>,
    pub dogstatsd_host: Option<String>,
    pub dogstatsd_port: Option<u16>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_config() {
        let raw =
            "{\"physical_topic_name\": \"test\", \"logical_topic_name\": \"test\", \"broker_config\": {\"bootstrap.servers\": \"127.0.0.1:9092\", \"queued.max.messages.kbytes\": 10000}}";

        let topic_config: TopicConfig = serde_json::from_str(raw).unwrap();

        assert_eq!(
            topic_config.broker_config["queued.max.messages.kbytes"],
            "10000"
        );
    }
}
