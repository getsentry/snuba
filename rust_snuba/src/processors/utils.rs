use serde::{Deserializer, Deserialize};

pub const DEFAULT_RETENTION_DAYS: u16 = 90;

pub fn default_retention_days() -> Option<u16> {
    Some(DEFAULT_RETENTION_DAYS)
}

pub fn hex_to_u64<'de, D>(deserializer: D) -> Result<u64, D::Error>
where
    D: Deserializer<'de>,
{
    let hex = String::deserialize(deserializer)?;
    u64::from_str_radix(&hex, 16).map_err(serde::de::Error::custom)
}
