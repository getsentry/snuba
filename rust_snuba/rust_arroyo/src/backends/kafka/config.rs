use rdkafka::config::ClientConfig as RdKafkaConfig;
use std::collections::HashMap;

#[derive(Debug, Clone)]
pub struct KafkaConfig {
    config_map: HashMap<String, String>,
}

impl KafkaConfig {
    pub fn new_config(
        bootstrap_servers: Vec<String>,
        override_params: Option<HashMap<String, String>>,
    ) -> Self {
        let mut config_map = HashMap::new();
        config_map.insert("bootstrap.servers".to_string(), bootstrap_servers.join(","));

        let config = Self { config_map };

        apply_override_params(config, override_params)
    }

    pub fn new_consumer_config(
        bootstrap_servers: Vec<String>,
        group_id: String,
        auto_offset_reset: String,
        _strict_offset_reset: bool, // TODO: Implement this
        override_params: Option<HashMap<String, String>>,
    ) -> Self {
        let mut config = KafkaConfig::new_config(bootstrap_servers, None);
        config.config_map.insert("group.id".to_string(), group_id);
        config
            .config_map
            .insert("enable.auto.commit".to_string(), "false".to_string());
        config
            .config_map
            .insert("auto.offset.reset".to_string(), auto_offset_reset);

        apply_override_params(config, override_params)
    }

    pub fn new_producer_config(
        bootstrap_servers: Vec<String>,
        override_params: Option<HashMap<String, String>>,
    ) -> Self {
        let config = KafkaConfig::new_config(bootstrap_servers, None);

        apply_override_params(config, override_params)
    }
}

impl From<KafkaConfig> for RdKafkaConfig {
    fn from(item: KafkaConfig) -> Self {
        let mut config_obj = RdKafkaConfig::new();
        for (key, val) in item.config_map.iter() {
            config_obj.set(key, val);
        }
        config_obj
    }
}

fn apply_override_params(
    mut config: KafkaConfig,
    override_params: Option<HashMap<String, String>>,
) -> KafkaConfig {
    match override_params {
        Some(params) => {
            for (param, value) in params {
                config.config_map.insert(param, value);
            }
        }
        None => {}
    }
    config
}

#[cfg(test)]
mod tests {
    use super::KafkaConfig;
    use rdkafka::config::ClientConfig as RdKafkaConfig;
    use std::collections::HashMap;

    #[test]
    fn test_build_consumer_configuration() {
        let config = KafkaConfig::new_consumer_config(
            vec!["localhost:9092".to_string()],
            "my-group".to_string(),
            "error".to_string(),
            false,
            Some(HashMap::from([(
                "queued.max.messages.kbytes".to_string(),
                "1000000".to_string(),
            )])),
        );

        let rdkafka_config: RdKafkaConfig = config.into();
        assert_eq!(
            rdkafka_config.get("queued.max.messages.kbytes"),
            Some("1000000")
        );
    }
}
