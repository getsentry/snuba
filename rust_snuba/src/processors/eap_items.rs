use anyhow::Context;
use chrono::DateTime;
use seq_macro::seq;
use serde::Serialize;
use std::collections::HashMap;
use uuid::Uuid;

use prost::Message;
use sentry_arroyo::backends::kafka::types::KafkaPayload;
use sentry_protos::snuba::v1::any_value::Value;
use sentry_protos::snuba::v1::TraceItem;

use crate::config::ProcessorConfig;
use crate::processors::utils::enforce_retention;
use crate::types::{InsertBatch, KafkaMessageMetadata};

pub fn process_message(
    msg: KafkaPayload,
    _metadata: KafkaMessageMetadata,
    config: &ProcessorConfig,
) -> anyhow::Result<InsertBatch> {
    let payload: &[u8] = msg.payload().context("Expected payload")?;
    let trace_item = TraceItem::decode(payload)?;
    let origin_timestamp = DateTime::from_timestamp(trace_item.received.unwrap().seconds, 0);
    let retention_days = Some(enforce_retention(
        Some(trace_item.retention_days as u16),
        &config.env_config,
    ));
    let mut eap_item = EAPItem::try_from(trace_item)?;

    eap_item.retention_days = retention_days;

    InsertBatch::from_rows([eap_item], origin_timestamp)
}

#[derive(Debug, Default, Serialize)]
struct EAPItem {
    organization_id: u64,
    project_id: u64,
    item_type: u8,
    timestamp: u32,
    trace_id: Uuid,
    item_id: u128,

    #[serde(flatten)]
    attributes: AttributeMap,

    sampling_factor: f64,
    sampling_weight: u64,

    retention_days: Option<u16>,
}

impl TryFrom<TraceItem> for EAPItem {
    type Error = anyhow::Error;

    fn try_from(from: TraceItem) -> Result<Self, Self::Error> {
        let timestamp = from.timestamp.context("Expected a timestamp")?;
        let mut eap_item = EAPItem {
            organization_id: from.organization_id,
            project_id: from.project_id,
            item_type: from.item_type as u8,
            trace_id: Uuid::parse_str(&from.trace_id)?,
            item_id: read_item_id(from.item_id),
            timestamp: timestamp.seconds as u32,
            attributes: Default::default(),
            retention_days: Default::default(),
            sampling_factor: 1.0,
            sampling_weight: 1,
        };

        for (key, value) in from.attributes {
            match value.value {
                Some(Value::StringValue(string)) => eap_item.attributes.insert_string(key, string),
                Some(Value::DoubleValue(double)) => eap_item.attributes.insert_float(key, double),
                Some(Value::IntValue(int)) => eap_item.attributes.insert_int(key, int),
                Some(Value::BoolValue(bool)) => eap_item.attributes.insert_bool(key, bool),
                Some(Value::BytesValue(_)) => (),
                Some(Value::ArrayValue(_)) => (),
                Some(Value::KvlistValue(_)) => (),
                None => (),
            }
        }

        if from.client_sample_rate > 0.0 {
            eap_item.sampling_factor *= from.client_sample_rate;
        }

        if from.server_sample_rate > 0.0 {
            eap_item.sampling_factor *= from.server_sample_rate;
        }

        // Lower precision to compensate floating point errors.
        eap_item.sampling_weight = (1.0 / eap_item.sampling_factor).round() as u64;
        eap_item.sampling_factor = (eap_item.sampling_factor * 1e9).round() / 1e9;

        Ok(eap_item)
    }
}

fn fnv_1a(input: &[u8]) -> u32 {
    const FNV_1A_PRIME: u32 = 16777619;
    const FNV_1A_OFFSET_BASIS: u32 = 2166136261;

    let mut res = FNV_1A_OFFSET_BASIS;
    for byt in input {
        res ^= *byt as u32;
        res = res.wrapping_mul(FNV_1A_PRIME);
    }

    res
}

fn read_item_id(from: Vec<u8>) -> u128 {
    let (item_id_bytes, _) = from.split_at(std::mem::size_of::<u128>());
    u128::from_le_bytes(item_id_bytes.try_into().unwrap())
}

macro_rules! seq_attrs {
    ($($tt:tt)*) => {
        seq!(N in 0..40 {
            $($tt)*
        });
    }
}

seq_attrs! {
#[derive(Debug, Default, Serialize)]
pub(crate) struct AttributeMap {
    attributes_bool: HashMap<String, bool>,
    attributes_int: HashMap<String, i64>,
    #(
    #[serde(skip_serializing_if = "HashMap::is_empty")]
    attributes_string_~N: HashMap<String, String>,

    #[serde(skip_serializing_if = "HashMap::is_empty")]
    attributes_float_~N: HashMap<String, f64>,
    )*
}
}

impl AttributeMap {
    pub fn insert_string(&mut self, k: String, v: String) {
        seq_attrs! {
            let attr_str_buckets = [
                #(
                &mut self.attributes_string_~N,
                )*
            ];
        };

        attr_str_buckets[(fnv_1a(k.as_bytes()) as usize) % attr_str_buckets.len()].insert(k, v);
    }

    pub fn insert_float(&mut self, k: String, v: f64) {
        seq_attrs! {
            let attr_num_buckets = [
                #(
                &mut self.attributes_float_~N,
                )*
            ];
        }

        attr_num_buckets[(fnv_1a(k.as_bytes()) as usize) % attr_num_buckets.len()].insert(k, v);
    }

    pub fn insert_bool(&mut self, k: String, v: bool) {
        // double write as float and bool
        self.insert_float(k.clone(), v as u8 as f64);
        self.attributes_bool.insert(k, v);
    }

    pub fn insert_int(&mut self, k: String, v: i64) {
        // double write as float and int
        self.insert_float(k.clone(), v as f64);
        self.attributes_int.insert(k, v);
    }
}

#[cfg(test)]
mod tests {
    use prost_types::Timestamp;
    use sentry_protos::snuba::v1::TraceItemType;

    use super::*;

    #[test]
    fn test_fnv_1a() {
        assert_eq!(fnv_1a("test".as_bytes()), 2949673445)
    }

    #[test]
    fn test_item_id_is_properly_decoded() {
        let item_id = Uuid::new_v4();
        let trace_item = TraceItem {
            attributes: Default::default(),
            item_id: item_id.as_u128().to_le_bytes().to_vec(),
            item_type: TraceItemType::Span.into(),
            organization_id: 1,
            project_id: 1,
            received: Default::default(),
            retention_days: 1,
            timestamp: Some(Timestamp {
                seconds: 1745562493,
                nanos: 0,
            }),
            trace_id: Uuid::new_v4().to_string(),
            client_sample_rate: 1.0,
            server_sample_rate: 1.0,
        };
        let eap_item = EAPItem::try_from(trace_item);

        assert!(eap_item.is_ok());
        assert_eq!(item_id.as_u128(), eap_item.unwrap().item_id);
    }
}
