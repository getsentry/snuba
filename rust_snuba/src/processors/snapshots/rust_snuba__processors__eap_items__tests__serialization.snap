---
source: src/processors/eap_items.rs
expression: item
---
{
  "attributes_bool": {
    "my.false.bool.field": false,
    "my.true.bool.field": true,
    "sentry.is_segment": true
  },
  "attributes_float_1": {
    "my.true.bool.field": 1.0
  },
  "attributes_float_14": {
    "my.int.field": 2000.0
  },
  "attributes_float_17": {
    "my.float.field": 101.2
  },
  "attributes_float_21": {
    "my.neg.float.field": -101.2
  },
  "attributes_float_25": {
    "sentry.exclusive_time_ms": 0.228,
    "sentry.start_timestamp_precise": 1721319572.616648
  },
  "attributes_float_26": {
    "sentry.is_segment": 1.0
  },
  "attributes_float_31": {
    "num_of_spans": 50.0
  },
  "attributes_float_4": {
    "sentry.end_timestamp_precise": 1721319572.768806
  },
  "attributes_float_5": {
    "my.neg.field": -100.0,
    "sentry.duration_ms": 152.0
  },
  "attributes_float_6": {
    "my.false.bool.field": 0.0
  },
  "attributes_float_7": {
    "sentry.received": 1721319572.877828
  },
  "attributes_int": {
    "my.int.field": 2000,
    "my.neg.field": -100,
    "sentry.duration_ms": 152
  },
  "attributes_string_0": {
    "relay_protocol_version": "3"
  },
  "attributes_string_11": {
    "sentry.segment_id": "8873a98879faf06d",
    "sentry.transaction.method": "POST"
  },
  "attributes_string_12": {
    "sentry.event_id": "d826225de75d42d6b2f01b957d51f18f",
    "server_name": "D23CXQ4GK2.local"
  },
  "attributes_string_14": {
    "sentry.status": "ok"
  },
  "attributes_string_17": {
    "relay_endpoint_version": "3",
    "relay_no_cache": "False",
    "relay_use_post_or_schedule": "True",
    "spans_over_limit": "False"
  },
  "attributes_string_18": {
    "sentry.segment.name": "/api/0/relays/projectconfigs/",
    "sentry.status_code": "200"
  },
  "attributes_string_19": {
    "sentry.op": "http.server"
  },
  "attributes_string_20": {
    "sentry.transaction": "/api/0/relays/projectconfigs/"
  },
  "attributes_string_21": {
    "sentry.thread.name": "uWSGIWorker1Core0"
  },
  "attributes_string_23": {
    "sentry.profile_id": "56c7d1401ea14ad7b4ac86de46baebae"
  },
  "attributes_string_24": {
    "thread.id": "8522009600"
  },
  "attributes_string_25": {
    "http.status_code": "200",
    "sentry.release": "backend@24.7.0.dev0+c45b49caed1e5fcbf70097ab3f434b487c359b6b",
    "sentry.sdk.name": "sentry.python.django",
    "sentry.transaction.op": "http.server"
  },
  "attributes_string_26": {
    "relay_id": "88888888-4444-4444-8444-cccccccccccc"
  },
  "attributes_string_27": {
    "sentry.trace.status": "ok"
  },
  "attributes_string_28": {
    "sentry.category": "http"
  },
  "attributes_string_29": {
    "sentry.environment": "development"
  },
  "attributes_string_3": {
    "sentry.thread.id": "8522009600"
  },
  "attributes_string_30": {
    "sentry.sdk.version": "2.7.0"
  },
  "attributes_string_31": {
    "sentry.platform": "python",
    "sentry.user": "ip:127.0.0.1"
  },
  "attributes_string_32": {
    "relay_use_post_or_schedule_rejected": "version"
  },
  "attributes_string_38": {
    "sentry.description": "/api/0/relays/projectconfigs/"
  },
  "attributes_string_6": {
    "thread.name": "uWSGIWorker1Core0"
  },
  "item_id": 9832388815107059821,
  "item_type": 1,
  "organization_id": 1,
  "project_id": 1,
  "retention_days": 90,
  "sampling_weight": 50,
  "timestamp": 1721319572,
  "trace_id": "d099bf9a-d5a1-43cf-8f83-a98081d0ed3b"
}
